### PR TITLE
Fix UB in serpent

### PR DESCRIPTION
**Pull request by eagleoflqj**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**

UB in crypto_serpent.c: out of boundary memory read.
UB in crypto_serpent_algo.c: shift of int more than 31.
Partially solved #15849 and #17898 

**Test plan**

https://travis-ci.com/github/radareorg/radare2/jobs/427537252#L1047
https://travis-ci.com/github/eagleoflqj/radare2/jobs/430534124#L1020
db/tools/rahash2 fixed.

**Closing issues**

...

### CODE REVIEW

File: libr/crypto/p/crypto_serpent_algo.c [hunk 1]
```diff
@@ -77,7 +77,7 @@ void apply_IP(ut32 in[DW_BY_BLOCK], ut32 out[DW_BY_BLOCK]) {
 	for (i = 0; i < DW_BY_BLOCK*32; i++) {
 		index = IPTable[i];
 		out[i/32] ^= (-(ut32)get_bit (index%32, in[index/32]) ^ out[i/32])
-			& (1 << i);
+			& (1 << (i & 0x1f));
```
Comment (pelijah) [inline libr/crypto/p/crypto_serpent_algo.c:80]: Wouldn't it be better to use `i % 32`?

Comment (eagleoflqj) [reply libr/crypto/p/crypto_serpent_algo.c:80]: I don't know whether there's a compiler that won't do this conversion, so I used &0x1f directly. Just checked the assembly on rpi4 and yes, %32 is treated as &31 properly. Is there a readability concern?

File: libr/crypto/p/crypto_serpent_algo.c [hunk 2]
```diff
@@ -77,7 +77,7 @@ void apply_IP(ut32 in[DW_BY_BLOCK], ut32 out[DW_BY_BLOCK]) {
 	for (i = 0; i < DW_BY_BLOCK*32; i++) {
```
Comment (trufae) [inline libr/crypto/p/crypto_serpent_algo.c:77]: add a space around *

File: libr/crypto/p/crypto_serpent_algo.c (hunk 1 above)
Comment (trufae) [reply libr/crypto/p/crypto_serpent_algo.c:80]: this is a shift left, this means that using % will result on wrong behaviour because it will be using a rotated value instead of a shifted one.

The variable `i` goes from 0 to DW_BY_BLOCK * 32, which is 4.. this is.. this code will try to shift from 0 to 128 bits. this is a clear UB. and for me it seems like using % will result in better behaviour. but its different than the current one. I'm not sure if we have tests for this or if you tried to compare this with other implementations of serpent. but if we are just xoring the buffer aganist 0s.. its probably better to break before doing useless ops..

Can you add a test and compare the result with another implementation. https://github.com/sysopfb/pyserpent this one seems like a good candidate for messing up and comparing outputs.

File: libr/crypto/p/crypto_serpent.c [hunk 3]
```diff
@@ -50,11 +50,14 @@ static bool update(RCrypto *cry, const ut8 *buf, int len) {
...
 	// Construct ut32 blocks from byte stream
-	for (j = 0; j < size/4; j++) {
-		ibuf[j] = r_read_le32(&buf[4*j]);
+	for (j = 0; j < len/4; j++) {
+		ibuf[j] = r_read_le32 (&buf[4*j]);
+	}
+	if (len & 0x3) {
+		ut8 tail[4] = {0}; // Zero padding
+		memcpy (tail, buf + (len & ~0x3), len & 0x3);
+		ibuf[len/4] = r_read_le32 (tail);
```
Comment (trufae) [inline libr/crypto/p/crypto_serpent.c:59]: add spaces around /, * in all the file.

also &buf[4*j] is the same as buf + (4*j) , which is imho cleaner and easy to read

**Review by trufae — CHANGES_REQUESTED** [review]

**Comment by eagleoflqj** [comment]
The implementation you give is python2 (I have limited understanding), and seems only supports CBC mode (not ECB here). So I tried the [full submission package](http://www.cl.cam.ac.uk/~rja14/Papers/serpent.tar.gz) in the [official website of serpent](https://www.cl.cam.ac.uk/~rja14/serpent.html) on ubuntu x64. It supports 256bit mode only and the input/output is given in big endian hex. I tested with cases in db/tools/rahash2 which is 256bit.
```sh
# after cd floppy1 and make
liumeo@liumeo-ubuntu:~/serpent/floppy1$ ./serpent-test -e -k 5657575757575757575757575757575757575757575757575757575757575757 -p 56575757575757575757575757575757
plainText=56575757575757575757575757575757
userKey=5657575757575757575757575757575757575757575757575757575757575757
cipherText=b37300fa7f43dfa0f2742e3dcf8855d3

liumeo@liumeo-ubuntu:~/serpent/floppy1$ ./serpent-test -d -k 48FF814810492C9500B3C5829F45D62B48FF814810492C9500B3C5829F45D62B -c 844D5CAD144702EA844D5CAD144702EA
cipherText=844d5cad144702ea844d5cad144702ea
userKey=48ff814810492c9500b3c5829f45d62b48ff814810492c9500b3c5829f45d62b
plainText=f28ab2d115b4d23f747180a0fa8d7c67

liumeo@liumeo-ubuntu:~/serpent/floppy1$ ./serpent-test -e -k fedcba9876543210fedcba9876543210fedcba9876543210fedcba9876543210 -p 00000000000000000000000000311210
plainText=00000000000000000000000000311210
userKey=fedcba9876543210fedcba9876543210fedcba9876543210fedcba9876543210
cipherText=fafe561b7389b081ea46818e51fc7d41
```
In both cases, official implementation produces the same result with our implementation. The third one is the case I add for testing zero padding.
And here's a python function for convenience.
```py
def switch_endian(s):
    r = ''
    n = len(s)
    for i in range(0, n, 2):
        r = s[i] + s[i + 1] + r
    return r

switch_endian("b37300fa7f43dfa0f2742e3dcf8855d3")
```
I think now our implementation is correct (at least for 256bit mode).

**Review by XVilka — APPROVED** [review]

**Review by trufae — APPROVED** [review]
